src/main/java/com/arpnetworking/metrics/mad/sources/TransformingSource.java

```diff
@@ -0,0 +1,490 @@
+ * and merges {@link Metric} instances within each {@link Record}
+ * together while, optionally, removing, injecting, and modifying dimensions
+ * and metrics if the name matches a regular expression with a new name generated
+ * through replacement of all matches in the original name.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ * @author Ville Koskela (ville dot koskela at inscopemetrics dot com)
+ */
+public final class TransformingSource extends BaseSource {
+
+    @Override
+    public void start() {
+        _source.start();
+    }
+
+    @Override
+    public void stop() {
+        _source.stop();
+    }
+
+    /**
+     * Generate a Steno log compatible representation.
+     *
+     * @return Steno log compatible representation.
+     */
+    @LogValue
+    public Object toLogValue() {
+        return LogValueMapFactory.builder(this)
+                .put("source", _source)
+                .put("tranformations", _transformations)
+                .build();
+    }
+
+    @Override
+    public String toString() {
+        return toLogValue().toString();
+    }
+
+    private TransformingSource(final Builder builder) {
+        super(builder);
+        _source = builder._source;
+        _transformations = builder._transformations;
+
+        _source.attach(new TransformingObserver(this, _transformations));
+    }
+
+    private final Source _source;
+    private final ImmutableList<TransformationSet> _transformations;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TransformingSource.class);
+    private static final Splitter.MapSplitter TAG_SPLITTER = Splitter.on(';').omitEmptyStrings().trimResults().withKeyValueSeparator('=');
+
+    // NOTE: Package private for testing
+    /* package private */ static final class TransformingObserver implements Observer {
+
+        /* package private */ TransformingObserver(
+                final TransformingSource source,
+                final ImmutableList<TransformationSet> transformations) {
+            _source = source;
+            _transformations = transformations;
+        }
+
+        @Override
+        public void notify(final Observable observable, final Object event) {
+            if (!(event instanceof Record)) {
+                LOGGER.error()
+                        .setMessage("Observed unsupported event")
+                        .addData("event", event)
+                        .log();
+                return;
+            }
+
+            // Merge the metrics in the record together
+            final Record record = (Record) event;
+            final Map<Key, Map<String, MergingMetric>> mergedMetrics = Maps.newHashMap();
+            final LinkedHashMap<String, Map<String, String>> variablesMap = new LinkedHashMap<>();
+            variablesMap.put("dimension", record.getDimensions());
+            variablesMap.put("env", System.getenv());
+
+            for (TransformationSet transformation : _transformations) {
```

Comment:
I'm not convinced that the transformations should be treated as a single context. As a single context there can be only a single set of dimensions. In the case of say CollectD this is not necessarily the case. Consider cpu extracting core into a dimension; that doesn't apply to memory (were they in the same UOW). I think each transformation needs to create its own UOW from the existing one with the matching metrics and the dimensions that are elected. If we want to join them back (assuming dimension sets match) that's a possible optimization. Thoughts?

Reply:
To elaborate:

- There is the incoming UOW
- Metrics matching a transformation are transformed into a new UOW which inherits its dimensions from the parent (post inject and remove) along with any injections/removals from the transform
- Any metrics included in a new UOW are removed from the parent UOW
- Metrics matching multiple UOWs are duplicated across them (subsequent changes adding separate matching rules to transformers can make this more elaborate)

Reply:
That's essentially what we're doing with the merge map.